{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Instruction\n",
    "### Installation of BLAST \n",
    "Install NCBI standalone BLAST from (https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download). \\\n",
    "Set the path to the BLAST.\n",
    "\n",
    "### Make database\n",
    "Download a fast file containing the transcriptome of your species. The example link is (http://useast.ensembl.org/info/data/ftp/index.html). \\\n",
    "The file is used to minimize the non-specific binding of oligonucleotides. \\\n",
    "Make a local database from the file according to (https://www.ncbi.nlm.nih.gov/books/NBK569841/). \\\n",
    "Example databases are included in ../data/ . Only database with ensemple transcript ID is accepted.\n",
    "\n",
    "### Download \n",
    "Download the fasta file of mRNA of your interest from NCBI.\n",
    "The example file is included in ../data/slc17a7/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# import sys\n",
    "# sys.path.append('../oligodesigner')\n",
    "\n",
    "from oligodesigner import blockParse_py3\n",
    "from oligodesigner import io\n",
    "from oligodesigner import parse\n",
    "from oligodesigner import sequence\n",
    "from oligodesigner import parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'oligodesigner.sequence' from '/home/tmurakami/src/split-oligo-designer/oligodesigner/sequence.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import importlib\n",
    "importlib.reload(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configs\n",
    "cdna_database = {\n",
    "                 'human': './data/human_transcriptome/human_transcriptome_db',\n",
    "                 'mouse': './data/mouse_transcriptome/mouse_transcriptome_db',\n",
    "                 'rat'  : './data/rat_transcriptome/rat_transcriptome_db'\n",
    "}\n",
    "\n",
    "\n",
    "# hcr_b_set = {\n",
    "#     'B1':{\n",
    "#         'splitseq_ll':'GAGGAGGGCAGCAAACGG','splitseq_rr':'GAAGAGTCTTCCTTTACG','anchorseq_ll':'aa','anchorseq_rr':'at',\n",
    "#         'splitseq_lr':'GGCAGCAAACGGGAAGAG','splitseq_rl':'GCATTCTTTCTTGAGGAG','anchorseq_lr':'tt','anchorseq_rl':'tt'\n",
    "#     }\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameter settings\n",
    "mFISH3D_param = {\n",
    "    'input_file':'/scratch1/tmurakami/dev_data/oligodesigner/slc17a7_hs_debug.fasta',#_debug\n",
    "    'hcr_bnum':'B1',\n",
    "    'cdna_database':cdna_database['human'],\n",
    "    'minimum_offtarget_gap':100,\n",
    "    # 'oligo_number': 150\n",
    "}\n",
    "\n",
    "oligominer_param = {\n",
    "    'l':20,\n",
    "    'L':20,\n",
    "    'gcPercent':25,\n",
    "    'GCPercent':75,\n",
    "    'tm':0,\n",
    "    'TM':100,\n",
    "    'X':'AAAAAA,TTTTTT,CCCCCC,GGGGGG',\n",
    "    'sal':390,#default value of blockParse\n",
    "    'form':30,\n",
    "    'sp':1,\n",
    "    'concA':25,#default value of blockParse\n",
    "    'concB':25,#default value of blockParse\n",
    "    'headerVal':None,\n",
    "    'bedVal':False,\n",
    "    'OverlapModeVal':False,\n",
    "    'verbocity':False,\n",
    "    'reportVal':True,\n",
    "    'debugVal':False,\n",
    "    'metaVal':False,\n",
    "    'outNameVal':None,\n",
    "    'nn_table':'DNA_NN3'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get transcripts that is the same gene product of the input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no refseq id was found\n",
      "The sequence is the most similar to SLC17A7 with the score of 0.6463361991574816\n",
      "Homologous transcript(s) was identified. This may impact on the selection of the oligos\n",
      "Homologous transcript SLC17A8 was found with the score of 0.1779060160120019\n"
     ]
    }
   ],
   "source": [
    "input_file = mFISH3D_param['input_file']\n",
    "transcriptomeDatabase = mFISH3D_param['cdna_database']\n",
    "\n",
    "self_remove = True \n",
    "# Default True. set self_remove True if you want to remove the input sequence from off-target analysis.\n",
    "# Otherwise, the gene of your interest could be regarded as an off-target product.\n",
    "sequence_homology_search = True\n",
    "# If True, the search of the homology will be performed regardless of the existence of refseq id in fasta.\n",
    "\n",
    "if self_remove:\n",
    "    seqid = sequence.check_refseqid_exist(input_file, return_entrezid=True)\n",
    "    if seqid is None:\n",
    "        seqid = sequence.get_homology_in_database(input_file, transcriptomeDatabase, num_threads=32)\n",
    "    elif sequence_homology_search:\n",
    "        _ = sequence.get_homology_in_database(input_file, transcriptomeDatabase, num_threads=32)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "hcr_initiator_set = parameters.hcr_b_set[mFISH3D_param['hcr_bnum']]\n",
    "minimum_offtarget_gap = mFISH3D_param['minimum_offtarget_gap']\n",
    "# oligo_number = mFISH3D_param['oligo_number']\n",
    "\n",
    "# Make path to temp files and result file\n",
    "(oligominer_fastq, oligominer_fasta, blast_result, oligo_sets, param_file) = io.gen_file_path(input_file)\n",
    "# Save parameters to csv file.\n",
    "io.record_param(mFISH3D_param, oligominer_param, param_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 of 2949\n",
      "132 candidate probes identified in 2.94 kb yielding 44.96 candidates/kb\n"
     ]
    }
   ],
   "source": [
    "# Run OligoMiner\n",
    "blockParse_py3.runSequenceCrawler(input_file, *parse.oligominer_param_parser(oligominer_param))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# converting fastq to fasta\n",
    "io.convert_fastq2fasta(oligominer_fastq,oligominer_fasta)\n",
    "io.add_id(oligominer_fasta)\n",
    "\n",
    "# run blast\n",
    "sequence.run_blast(oligominer_fasta,blast_result,transcriptomeDatabase)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read blast result file, exclude sequence which is homologous to the gene of interest from result\n",
    "df_blast = sequence.exclude_self(io.read_blast(blast_result), seqid)\n",
    "\n",
    "# Remove combination if sstart and send are close enough and if they are rective combination.\n",
    "oligo_to_be_removed = sequence.get_off_targeting_oligo(minimum_offtarget_gap, df_blast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "oligominer_df = io.dataframe_from_oligominer(oligominer_fastq)\n",
    "selected_oligo_df = sequence.remove_oligo(oligominer_df, oligo_to_be_removed)\n",
    "\n",
    "interval = sequence.calc_oligo_interval(selected_oligo_df)\n",
    "selected_oligo_df = pd.concat([selected_oligo_df,interval],axis=1)\n",
    "\n",
    "# Make oligos with HCR sequences.\n",
    "res = sequence.add_hcr_seq_v3(selected_oligo_df, hcr_initiator_set)\n",
    "res.to_csv(oligo_sets, sep=',', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>interval_after</th>\n",
       "      <th>mean_interval</th>\n",
       "      <th>hcr_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_foomoo:1-20</td>\n",
       "      <td>NM_foomoo:1-20_0</td>\n",
       "      <td>(A, C, T, T, G, C, A, G, C, C, T, C, C, T, T, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>34.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_foomoo:54-73</td>\n",
       "      <td>NM_foomoo:54-73_1</td>\n",
       "      <td>(T, G, G, A, C, C, C, C, G, G, G, A, A, C, C, ...</td>\n",
       "      <td>54</td>\n",
       "      <td>73</td>\n",
       "      <td>59.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_foomoo:132-151</td>\n",
       "      <td>NM_foomoo:132-151_2</td>\n",
       "      <td>(C, G, G, C, A, G, G, A, G, C, C, G, C, C, A, ...</td>\n",
       "      <td>132</td>\n",
       "      <td>151</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_foomoo:153-172</td>\n",
       "      <td>NM_foomoo:153-172_3</td>\n",
       "      <td>(A, G, T, T, C, C, G, C, C, A, G, G, A, G, G, ...</td>\n",
       "      <td>153</td>\n",
       "      <td>172</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_foomoo:174-193</td>\n",
       "      <td>NM_foomoo:174-193_4</td>\n",
       "      <td>(G, G, A, A, G, C, T, A, G, C, G, G, G, T, C, ...</td>\n",
       "      <td>174</td>\n",
       "      <td>193</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>NM_foomoo:2798-2817</td>\n",
       "      <td>NM_foomoo:2798-2817_126</td>\n",
       "      <td>(A, C, C, T, A, T, A, G, A, G, A, A, G, T, C, ...</td>\n",
       "      <td>2798</td>\n",
       "      <td>2817</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>NM_foomoo:2819-2838</td>\n",
       "      <td>NM_foomoo:2819-2838_127</td>\n",
       "      <td>(T, C, C, C, A, T, G, A, A, A, G, G, G, A, A, ...</td>\n",
       "      <td>2819</td>\n",
       "      <td>2838</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>NM_foomoo:2855-2874</td>\n",
       "      <td>NM_foomoo:2855-2874_128</td>\n",
       "      <td>(T, T, T, T, T, C, C, C, A, G, T, G, C, A, G, ...</td>\n",
       "      <td>2855</td>\n",
       "      <td>2874</td>\n",
       "      <td>23.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>NM_foomoo:2897-2916</td>\n",
       "      <td>NM_foomoo:2897-2916_130</td>\n",
       "      <td>(T, G, A, C, T, G, T, A, A, A, T, C, C, T, G, ...</td>\n",
       "      <td>2897</td>\n",
       "      <td>2916</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>NM_foomoo:2918-2937</td>\n",
       "      <td>NM_foomoo:2918-2937_131</td>\n",
       "      <td>(C, C, C, A, C, T, C, C, C, A, A, T, A, A, A, ...</td>\n",
       "      <td>2918</td>\n",
       "      <td>2937</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene                 oligo_id  \\\n",
       "0         NM_foomoo:1-20         NM_foomoo:1-20_0   \n",
       "1        NM_foomoo:54-73        NM_foomoo:54-73_1   \n",
       "2      NM_foomoo:132-151      NM_foomoo:132-151_2   \n",
       "3      NM_foomoo:153-172      NM_foomoo:153-172_3   \n",
       "4      NM_foomoo:174-193      NM_foomoo:174-193_4   \n",
       "..                   ...                      ...   \n",
       "124  NM_foomoo:2798-2817  NM_foomoo:2798-2817_126   \n",
       "125  NM_foomoo:2819-2838  NM_foomoo:2819-2838_127   \n",
       "126  NM_foomoo:2855-2874  NM_foomoo:2855-2874_128   \n",
       "127  NM_foomoo:2897-2916  NM_foomoo:2897-2916_130   \n",
       "128  NM_foomoo:2918-2937  NM_foomoo:2918-2937_131   \n",
       "\n",
       "                                                   seq  start   end  \\\n",
       "0    (A, C, T, T, G, C, A, G, C, C, T, C, C, T, T, ...      1    20   \n",
       "1    (T, G, G, A, C, C, C, C, G, G, G, A, A, C, C, ...     54    73   \n",
       "2    (C, G, G, C, A, G, G, A, G, C, C, G, C, C, A, ...    132   151   \n",
       "3    (A, G, T, T, C, C, G, C, C, A, G, G, A, G, G, ...    153   172   \n",
       "4    (G, G, A, A, G, C, T, A, G, C, G, G, G, T, C, ...    174   193   \n",
       "..                                                 ...    ...   ...   \n",
       "124  (A, C, C, T, A, T, A, G, A, G, A, A, G, T, C, ...   2798  2817   \n",
       "125  (T, C, C, C, A, T, G, A, A, A, G, G, G, A, A, ...   2819  2838   \n",
       "126  (T, T, T, T, T, C, C, C, A, G, T, G, C, A, G, ...   2855  2874   \n",
       "127  (T, G, A, C, T, G, T, A, A, A, T, C, C, T, G, ...   2897  2916   \n",
       "128  (C, C, C, A, C, T, C, C, C, A, A, T, A, A, A, ...   2918  2937   \n",
       "\n",
       "     interval_after  mean_interval  \\\n",
       "0              34.0            NaN   \n",
       "1              59.0            NaN   \n",
       "2               2.0            NaN   \n",
       "3               2.0            NaN   \n",
       "4               2.0            NaN   \n",
       "..              ...            ...   \n",
       "124             2.0            NaN   \n",
       "125            17.0            NaN   \n",
       "126            23.0            NaN   \n",
       "127             2.0            NaN   \n",
       "128             NaN            NaN   \n",
       "\n",
       "                                               hcr_seq  \n",
       "0    (G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...  \n",
       "1    (G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...  \n",
       "2    (G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...  \n",
       "3    (G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...  \n",
       "4    (G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...  \n",
       "..                                                 ...  \n",
       "124  (G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...  \n",
       "125  (G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...  \n",
       "126  (G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...  \n",
       "127  (G, C, A, T, T, C, T, T, T, C, T, T, G, A, G, ...  \n",
       "128  (G, A, G, G, A, G, G, G, C, A, G, C, A, A, A, ...  \n",
       "\n",
       "[129 rows x 8 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "oligo",
   "language": "python",
   "name": "oligo"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
